src/moonlink/src/storage/mooncake_table/snapshot_read_output.rs

```diff
@@ -71,48 +75,120 @@ impl ReadOutput {
             .unwrap();
     }
 
+    fn separate_files_by_type(
+        data_file_paths: Vec<DataFileForRead>,
+    ) -> (TempFilesWithIndex, RemoteFilesWithIndex) {
+        let mut temp_files_with_index = Vec::new();
+        let mut remote_files_with_index = Vec::new();
+        for (index, cur_data_file) in data_file_paths.into_iter().enumerate() {
+            match cur_data_file {
+                DataFileForRead::TemporaryDataFile(file) => {
+                    temp_files_with_index.push((index, file));
+                }
+                DataFileForRead::RemoteFilePath((file_id, remote_filepath)) => {
+                    remote_files_with_index.push((index, file_id, remote_filepath));
+                }
+            }
+        }
+        (temp_files_with_index, remote_files_with_index)
+    }
+
+    fn process_temporary_files(
+        temp_files_with_index: Vec<(usize, String)>,
+        resolved_data_files: &mut [String],
+    ) {
+        for (index, file) in temp_files_with_index {
+            resolved_data_files[index] = file;
+        }
+    }
+    /// Processes remote files by downloading them in parallel and updating the cache.
+    /// This function ensures that the order of files is maintained while handling
+    /// cache operations and potential errors.
+    async fn process_remote_files(
+        &mut self,
+        object_storage_cache: Arc<dyn CacheTrait>,
+        filesystem_accessor: Arc<dyn BaseFileSystemAccess>,
+        remote_files_with_index: Vec<(usize, TableUniqueFileId, String)>,
+        resolved_data_files: &mut [String],
+        cache_handles: &mut Vec<NonEvictableHandle>,
+    ) -> Result<()> {
+        let futures = remote_files_with_index
+            .into_iter()
```

Comment:
I think it's better to apply upper limit on parallel copy, otherwise we have no control of resource (i.e., memory).
To me, a hard-coded heuristic cap like 128 looks good enough to me